substrate/frame/election-provider-multi-block/src/unsigned/miner.rs

```diff
@@ -0,0 +1,1979 @@
+			BoundedVec::with_bounded_capacity(pages as usize);
+		paged_assignments.bounded_resize(pages as usize, Default::default());
+		for assignment in trimmed_assignments {
+			// NOTE: this `page` index is LOCAL. It does not correspond to the actual page index of
+			// the snapshot map, but rather the index in the `voter_pages`.
+			let page = voter_page_fn(&assignment.who).ok_or(MinerError::InvalidPage)?;
+			let assignment_page =
+				paged_assignments.get_mut(page as usize).ok_or(MinerError::InvalidPage)?;
+			assignment_page.push(assignment);
+		}
+
+		// convert each page to a compact struct
+		let solution_pages: BoundedVec<SolutionOf<T>, T::Pages> = paged_assignments
+			.into_iter()
+			.enumerate()
+			.map(|(page_index, assignment_page)| {
+				// get the page of the snapshot that corresponds to this page of the assignments.
+				let page: PageIndex = page_index.saturated_into();
+				let voter_snapshot_page = voter_pages
+					.get(page as usize)
+					.ok_or(MinerError::SnapshotUnAvailable(SnapshotType::Voters(page)))?;
+
+				let voter_index_fn = {
+					let cache = helpers::generate_voter_cache::<T, _>(&voter_snapshot_page);
+					helpers::voter_index_fn_owned::<T>(cache)
+				};
+				<SolutionOf<T>>::from_assignment(
+					&assignment_page,
+					&voter_index_fn,
+					&target_index_fn,
+				)
+				.map_err::<MinerError<T>, _>(Into::into)
+			})
+			.collect::<Result<Vec<_>, _>>()?
+			.try_into()
+			.expect("`paged_assignments` is bound by `T::Pages`; length cannot change in iter chain; qed");
+
+		// now do the weight and length trim.
+		let mut solution_pages_unbounded = solution_pages.into_inner();
+		let _trim_length_weight =
+			Self::maybe_trim_weight_and_len(&mut solution_pages_unbounded, &voter_pages)?;
+		let solution_pages = solution_pages_unbounded
+			.try_into()
+			.expect("maybe_trim_weight_and_len cannot increase the length of its input; qed.");
+		miner_log!(debug, "trimmed {} voters due to length restriction.", _trim_length_weight);
+
+		// finally, wrap everything up. Assign a fake score here, since we might need to re-compute
+		// it.
+		let mut paged = PagedRawSolution { round, solution_pages, score: Default::default() };
+
+		// OPTIMIZATION: we do feasibility_check inside `compute_score`, and once later
+		// pre_dispatch. I think it is fine, but maybe we can improve it.
+		let score = Self::compute_score(&paged, &voter_pages, &all_targets, desired_targets)
+			.map_err::<MinerError<T>, _>(Into::into)?;
+		paged.score = score.clone();
+
+		miner_log!(
+			info,
+			"mined a solution with {} pages, score {:?}, {} winners, {} voters, {} edges, and {} bytes",
+			pages,
+			score,
+			paged.winner_count_single_page_target_snapshot(),
+			paged.voter_count(),
+			paged.edge_count(),
+			paged.using_encoded(|b| b.len())
+		);
+
+		Ok(paged)
+	}
+
+	/// perform the feasibility check on all pages of a solution, returning `Ok(())` if all good and
```

Comment:
The return type isn't `Ok(())`, old comment or that the code needs be adjusted?

Reply:
Yeah prior to merge I will go over the docs one last time and try and update the important ones, it has indeed slipped passed me and I had to update a lot of things quickly, but didn't update the docs 🙈